feature: provide cmake find_package configuration

elfin-sbreuers opened the pull request.

redboltz commented:
Thank you for sending the PR.
After I merged #147 , I've checked some other project such as https://github.com/msgpack/msgpack-c/blob/master/CMakeLists.txt

I noticed that CMakeLists.txt in the include directory is unnatural. I think that we can remove it and add install operation to root CMakeLists.txt. Could you update the PR?

In addition,
https://github.com/redboltz/mqtt_cpp/pull/149/files#diff-46193e2e675adcde81f1d095d8b3e1dbR7
doesn't seem good. `-lboost_system`  is UNIX specific notation. Maybe you need to use CMakeVariables something like `${Boost_SYSTEM_LIBRARY}`. See https://github.com/msgpack/msgpack-c/blob/master/example/cpp03/CMakeLists.txt#L97

redboltz commented:
@elfin-sbreuers , I got confused that #147 is your pull request. But it is by @marcel-behlau-elfin.
Sorry for the mix-up.

Anyway, I'd like to remove `include/CMakeLists.txt` and move the same functionality to `../CMakeLists.txt`.

I'm not expert of cmake. If placing `include/CMakeLists.txt` is better way than implement the functionality in `../CMakeLists.txt`, please let me know.

redboltz commented:
See the comment https://github.com/redboltz/mqtt_cpp/pull/147#issuecomment-448610916

@marcel-behlau-elfin , could you review the PR? I'm not sure how to provide cmake find_package.

ogis-kondo commented on include/CMakeLists.txt:
```diff
@@ -1,11 +1,20 @@
+add_library(${PROJECT_NAME} INTERFACE)
 
-target_include_directories(mqtt_cpp_iface INTERFACE ${HDR_ROOT} ${HDR_MQTT})
+target_link_libraries(${PROJECT_NAME} INTERFACE ${Boost_SYSTEM_LIBRARIES} -lpthread)
```
I think that `${Boost_SYSTEM_LIBRARIES}` isn't defined.
According to the document https://cmake.org/cmake/help/v3.8/module/FindBoost.html

```
Boost_<C>_LIBRARY      - Libraries to link for component <C> (may include
                         target_link_libraries debug/optimized keywords)
```

So you need to use `${Boost_SYSTEM_LIBRARY}`


`-lpthread` is platform dependent.

See
https://github.com/redboltz/mqtt_cpp/blob/master/CMakeLists.txt#L42
https://github.com/redboltz/mqtt_cpp/blob/master/test/CMakeLists.txt#L36

redboltz replied:
I think that `${Boost_SYSTEM_LIBRARIES}` isn't defined.
According to the document https://cmake.org/cmake/help/v3.8/module/FindBoost.html

```
Boost_<C>_LIBRARY      - Libraries to link for component <C> (may include
                         target_link_libraries debug/optimized keywords)
```

So you need to use `${Boost_SYSTEM_LIBRARY}`


`-lpthread` is platform dependent.

See
https://github.com/redboltz/mqtt_cpp/blob/master/CMakeLists.txt#L42
https://github.com/redboltz/mqtt_cpp/blob/master/test/CMakeLists.txt#L36

redboltz commented:
I don't know much about cmake find_package mechanism.
Could you tell me how your modification works?
Why `target_link_libraries` is required? `mqtt_cpp` is a header only library. Is it for resolve dependency?

elfin-sbreuers commented:
Hello,

Thanks for your valuable input. I updated the `CMakeLists.txt` according to your recommandations and added a Config.cmake file that allows for a proper usage in modern `cmake` applications. This includes the necessary link dependencies for threads and boost libraries.

You library is a header only library, but when using the boost system library, it is not header only, the system library needs to be added.

I did not move it up in the directory structure. I guess, one could argue about the right location of the installation procedure. I left it in `includes`, since the library is 'built' or located there and there should also be defined what is needed to properly build against the library.

Having said that, the inclusion and configuration of boost should probably go into the `test/CMakeList.txt`.

On a side note: apparently the usage of static boost libraries is only needed for your continuous-integration system. Your tests built just fine without it locally on my machine.
 
Are the current changes matching your requirements?

redboltz commented:
Thank you for update the PR. I need a time to study cmake mechanism. Please wait a couple of days(hopefully ;)

redboltz commented on README.md:
```diff
@@ -66,6 +66,17 @@ If you want to use MQTT on WebSocket, you need to define `MQTT_USE_WS` macro. mq
 
+## Usage in cmake project
+
+Add following lines to your `CMakeList.txt`
```
It should be `CmakeLists.txt`.

redboltz commented on README.md:
````diff
@@ -66,6 +66,17 @@ If you want to use MQTT on WebSocket, you need to define `MQTT_USE_WS` macro. mq
+
+```
+find_package(mqtt_cpp REQUIRED)
+target_link_libraries(${LIBNAME} LINK_PUBLIC mqtt_cpp::mqtt_cpp)
````
What does `${LIBNAME}` mean?
If if is a users target name, it should be something like `${YOUR_TARGET}`.

redboltz replied:
I've tested the PR on my environment.
I did `make install` mqtt_cpp. Then create the new cmake project.
The CMakeLists.txt is

```cmake
PROJECT (mytest)

find_package(mqtt_cpp REQUIRED)
target_link_libraries(${PROJECT} LINK_PUBLIC mqtt_cpp::mqtt_cpp)
```

I did

```sh
make build
cd build
cmake -DCMAKE_FIND_DEBUG_MODE=1 ..
```

Then I got the following error:

```
CMake Error at CMakeLists.txt:4 (target_link_libraries):
  Cannot specify link libraries for target "LINK_PUBLIC" which is not built
  by this project.
```

elfin-sbreuers replied:
Thanks for the input. I improved the readme. Apparently the target is missing that you add the `target_link_libraries` to. That means a target generation command like
```
add_library(${PROJECT_NAME} <dependencies>)
```
or
```
add_executable(${PROJECT_NAME} <dependencies>)
```
is missing.

elfin-sbreuers replied:
The `target_link_libraries` call would then also need the following call

```
target_link_libraries(${PROJECT_NAME} mqtt_cpp::mqtt_cpp)
```
since the `PROJECT` variable itself is not set.

redboltz replied:
I still got the error. 

I don't understand the following code at readme.

```cmake
set(LIBNAME "your_project_name")
add_library(${LIBNAME} <your dependencies here>)
find_package(mqtt_cpp REQUIRED)
target_link_libraries(${LIBNAME} LINK_PUBLIC mqtt_cpp::mqtt_cpp)
```

The user creates library? I think user creates application.

redboltz replied:
Could you give me complete and minimal example of `CMakeLists.txt` that uses installed `mqtt_cpp`?
I never success cmake, so far.

jonesmz replied:
Create a library that uses mqtt_cpp

```
PROJECT(my_mqtt)

MESSAGE(STATUS "Configuring ${PROJECT_NAME}")

FILE(GLOB SRC "*.cpp")

ADD_LIBRARY(my_mqtt ${SRC})
ADD_LIBRARY(my::mqtt ALIAS my_mqtt)

TARGET_INCLUDE_DIRECTORIES(my_mqtt PUBLIC ${MY_FRAMEWORKS_ROOT}/inc/)

TARGET_LINK_LIBRARIES(my_mqtt my::core my::log mqtt_cpp_iface Boost::system OpenSSL::SSL nlohmann_json::nlohmann_json )
```

Create an application that uses the library.

```
PROJECT(my-application)

MESSAGE(STATUS "Configuring ${PROJECT_NAME}")

FILE(GLOB SRC "*.cpp")

ADD_EXECUTABLE(my-application ${SRC})

TARGET_LINK_LIBRARIES(my-application my::mqtt Boost::program_options)

INSTALL(TARGETS my-application RUNTIME DESTINATION /usr/bin)
```

jonesmz replied:
What this PR does is make it so someone (e.g., a distribution packager for a linux distribution) can install mqtt_cpp via "make install", and them someone else who wants to use mqtt_cpp, in their CMakeLists.txt, can use the CMake "find_package" command to locate mqtt_cpp on the filesystem and use the code in their own application.

I imagine that this change will also make it easier to satisfy this issue: https://github.com/redboltz/mqtt_cpp/issues/77 (but that's just speculation on my part. I know nothing of vcpkg)

redboltz commented:
Could you rebase the PR from up to date master?

jonesmz commented:
@elfin-sbreuers I'm quite interested in seeing this work get merged.

Could you rebase your patch?

jonesmz commented on CMakeLists.txt:
```diff
@@ -35,8 +35,10 @@ ENDIF ()
+    SET (BOOSTVERSION "1.66.0")
 ELSE ()
     MESSAGE (STATUS "WebSocket disabled")
+    SET (BOOSTVERSION "1.59.0")
```
I'm not sure this is accurate anymore.

There were bigger changes that cause incompatibilities with versions of boost older than 1.67.

But if this does actually work, that'd be great to know about.

marcel-behlau-elfin replied:
Ohh, you are right. This will be reverted.
